lib/lpc55-rot-startup/src/images.rs

```diff
@@ -55,124 +38,360 @@ extern "C" {
-    }
+        // Because of our past experience with the implementation quality of the
+        // ROM, let's do some basic checks before handing it a blob to inspect,
+        // shall we?
 
-    pub fn get_image_version(&self) -> ImageVersion {
-        // SAFETY: We checked this previously
-        let header = unsafe { &*self.get_header() };
+        const MASK_WITHOUT_28: u32 = !(1 << 28);
+        let reset_vector = MASK_WITHOUT_28 & self.get_reset_vector();
 
-        ImageVersion {
-            epoch: header.epoch,
-            version: header.version,
+        // Verify that the reset vector is a valid Thumb-2 function pointer.
+        if reset_vector & 1 == 0 {
+            // This'll cause an immediate usage fault. Reject it.
+            return Err(ImageError::ResetVectorNotThumb2);
         }
-    }
 
-    fn pointer_range(&self) -> core::ops::Range<*const u8> {
-        let img_ptr = self.get_img_start() as *const u8;
-        // The MPU requires 32 byte alignment and so the compiler pads the
-        // image accordingly. The length field from the image header does not
-        // (and should not) account for this padding so we must do that here.
-        let img_size = (self.get_img_size().unwrap_lite() + 31) & !31;
+        // Ensure that the reset vector is within the runtime address range.
+        let runtime = self.run.start..(self.run.start + self.span.len() as u32);
+        if !runtime.contains(&reset_vector) {
+            return Err(ImageError::ResetVector);
+        }
+
+        // The image type is at offset 0x24. We only allow type 4.
+        //   - 0x0000 Normal image for unsecure boot
+        //   - 0x0001 Plain signed Image
+        //   - 0x0002 Plain CRC Image, CRC at offset 0x28
+        //   - 0x0004 Plain signed XIP Image
+        //   - 0x0005 Plain CRC XIP Image, CRC at offset 0x28
+        //   - 0x8001 Signed plain Image with KeyStore Included
+        if self.get_image_type() != 4 {
+            return Err(ImageError::UnsupportedType);
+        }
 
-        // Safety: this is unsafe because the pointer addition could overflow.
-        // If that happens, we'll produce an empty range or crash with a panic.
-        // We do not dereference these here pointers.
-        img_ptr..unsafe { img_ptr.add(img_size) }
+        Ok(())
     }
 
-    pub fn contains(&self, address: *const u8) -> bool {
-        self.pointer_range().contains(&address)
+    // Assuming a well-formed image, get the result of the ROM signature check
+    // routine.
+    fn check_signature(&self, syscon: &SYSCON) -> Result<(), ImageError> {
+        const HASHAES: u32 = 32 + 32 + 18;
+        const PMASK: u32 = 1 << (HASHAES % 32);
+        const REG_NUM: u32 = HASHAES / 32; // XXX must be 0, 1, or 2
+        match REG_NUM {
+            0 => clear_bit!(syscon.presetctrl0, PMASK),
+            1 => clear_bit!(syscon.presetctrl1, PMASK),
+            2 => clear_bit!(syscon.presetctrl2, PMASK),
+            _ => panic!(),
+        }
```

Comment:
Please switch to using the PAC APIs directly instead of setting the bits https://docs.rs/lpc55-pac/latest/lpc55_pac/syscon/presetctrl2/index.html